src/Interpreters/ClusterDiscovery.cpp

```diff
@@ -0,0 +1,465 @@
+#include <algorithm>
+#include <atomic>
+#include <chrono>
+#include <condition_variable>
+#include <mutex>
+#include <unordered_map>
+#include <unordered_set>
+
+#include <base/getFQDNOrHostName.h>
+#include <base/logger_useful.h>
+
+#include <Common/Exception.h>
+#include <Common/StringUtils/StringUtils.h>
+#include <Common/ZooKeeper/Types.h>
+#include <Common/setThreadName.h>
+
+#include <Core/ServerUUID.h>
+
+#include <Interpreters/Cluster.h>
+#include <Interpreters/ClusterDiscovery.h>
+#include <Interpreters/Context.h>
+
+#include <Poco/Exception.h>
+#include <Poco/JSON/JSON.h>
+#include <Poco/JSON/Object.h>
+#include <Poco/JSON/Parser.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+}
+
+namespace
+{
+
+fs::path getShardsListPath(const String & zk_root)
+{
+    return fs::path(zk_root + "/shards");
+}
+
+}
+
+/*
+ * Holds boolean flags for fixed set of keys.
+ * Flags can be concurrently set from different threads, and consumer can wait for it.
+ */
+template <typename T>
+class ClusterDiscovery::ConcurrentFlags
+{
+public:
+    template <typename It>
+    ConcurrentFlags(It begin, It end)
+    {
+        for (auto it = begin; it != end; ++it)
+            flags.emplace(*it, false);
+    }
+
+    void set(const T & key)
+    {
+        setFlag(key, true);
+        cv.notify_one();
+    }
+
+    /// `need_update` expected to be value from `flags` corresponding to some key
+    void set(std::atomic_bool & need_update)
+    {
+        any_need_update = true;
+        need_update = true;
+        cv.notify_one();
+    }
+
+    /// waits unit at least one flag is set
+    /// caller should handle all set flags (or set it again manually)
+    std::unordered_map<T, std::atomic_bool> & wait(std::atomic_bool & finished, std::chrono::milliseconds timeout)
+    {
+        std::unique_lock<std::mutex> lk(mu);
+        cv.wait_for(lk, timeout, [this, &finished]() -> bool { return any_need_update || finished; });
+
+        /// all set flags expected to be handled by caller
+        any_need_update = false;
+        return flags;
+    }
+
+    std::unique_lock<std::mutex> getLock() { return std::unique_lock<std::mutex>(mu); }
+
+    void notify() { cv.notify_one(); }
+
+private:
+    void setFlag(const T & key, bool value)
+    {
+        any_need_update = any_need_update || value;
+
+        auto it = flags.find(key);
+        if (it == flags.end())
+            throw DB::Exception(ErrorCodes::LOGICAL_ERROR, "Unknown value '{}'", key);
+        it->second = value;
+    }
+
+    std::condition_variable cv;
+    std::mutex mu;
```

Comment:
I would rather move cv and my out of this container

Reply:
And maybe also `any_need_update`.

Reply:
removed this method overload at all

Reply:
In that case there will not be reason to have `ConcurrentFlags` container at all. The main idea is to make caller not responsible for triggering condvar and setting `any_need_update`.

We can move `any_need_update` + `cv` + `mu` + `flags` (actually all fields of `ConcurrentFlags`) to main class directly. But it would be inconvenient to pass them all to callback for example https://github.com/vdimir/ClickHouse/blob/60ebda20465a3be3ee982205e072b8e00d22a984/src/Interpreters/ClusterDiscovery.cpp#L150

Reply:
Agree, but right now it looks like a leaky abstraction.

Reply:
Because of `getLock` and `notify()` ? It used only in shutdown, so maybe we can move `stop_flag` to `clusters_to_update` ?